test/sqllogictest/outer_join_simplification.slt

```diff
@@ -0,0 +1,317 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+mode cockroach
+
+statement ok
+create table foo_raw (a int4, b int8, u text)
+
+statement ok
+create materialized view foo as select * from foo_raw where a is not null and b is not null
+
+statement ok
+create table bar_raw (a int4, v text)
+
+statement ok
+create materialized view bar as select distinct on (a) a, v from bar_raw
+
+statement ok
+create materialized view ban_nn as select * from bar where a is not null
+
+statement ok
+create table baz_raw (b int8, c int2, w text)
+
+statement ok
+create materialized view baz as select distinct on (b) b, c, w from baz_raw where b is not null
+
+statement ok
+create table quux_raw (c int2, x text)
+
+statement ok
+create materialized view quux as select distinct on (c) c, x from quux_raw where c is not null
+
+# Demonstrate core semijoin idempotence simplification: semijoin removal.
```

Comment:
This is not really testing the new transform: If I comment out the new transform, the plan in this test doesn't change. From looking at it with the optimizer trace tool, it seems `redundant_join` and `join_fusion` are doing all the work here.

Reply:
It's "demonstrating" it. I figured that would be helpful. I'm not sure that removing it is helpful, nor debugging why `redundant_join` manages to be useful here but not generally. At least, not as part of this PR.

Reply:
Ok.

Reply:
> This is not really testing the new transform: If I comment out the new transform, the plan in this test doesn't change. From looking at it with the optimizer trace tool, it seems `redundant_join` and `join_fusion` are doing all the work here.

@ggevay can you create a follow-up issue for us to actually debug why "`redundant_join` manages to be useful here but not generally".

Reply:
Done: https://github.com/MaterializeInc/materialize/issues/16325